File: objective-c/sample.m
#import <assert.h>

#define NSLINKLIST_TEST

#ifdef NSLINKLIST_TEST
  #import "NSLinkList.h"
#endif

void
NSLinkListTest() {
  NSLinkList *list = [[NSLinkList alloc] init];
  assert(list);

  NSString *str1 = @"Hello, world.";
  NSString *str2 = @"Hope is a good thing.";
  NSString *str3 = @"Walking all the way taking you.";
  NSString *str4 = @"Let's get started.";

  [list pushFront:str1];
  [list insertAfter: [[list head] next] insertedObject:str2];
  [list insertBefore:[[list head] next] insertedObject:str3];

  // prints out
  [list traverse];
  [list foreach:@selector(extendContents:) withObject:str4];

  NSLog([list description]);

  [list removeAllNodes];

  NSLog([list description]);
}

int
main(int argc, const char *argv[]) {
  NSAutoreleasePool *pool = [[NSAutoreleasePool alloc] init];

#ifdef NSLINKLIST_TEST
  NSLog(@"-------- Begin NSLinkList tests --------");
  NSLinkListTest();
  NSLog(@"-------- End NSLinkList tests --------");
#endif

  // ends
  [pool drain];

  return 0;
}
/* vim:set cinoptions=>2 shiftwidth=2: */

File: cpp/sample.cpp
#include <string>

#define LINK_LIST_TEST

#ifdef LINK_LIST_TEST
    #include "link_list.h"
#endif

int
link_list_test()
{
        LinkNode<string> node = LinkNode<string>("hello");
        LinkList<string> list;

        const string str1("Hello, world.");
        const string str2("Hope is a good thing.");
        const string str3("Walking all the way taking you.");
        const string str4("Let's get started.");

        list.PushFront(str1);
        list.PushFront(str2);
        list.InsertBefore(list.GetHead()->GetNext(), str3);

        LinkList<string>::Foreach(&list, LinkNode<string>::Visit);
        list.PopFront();
        LinkList<string>::Foreach(&list, LinkNode<string>::Visit);
        list.PopBack();
        LinkList<string>::Foreach(&list, LinkNode<string>::Visit);

        return 1;
}

int main(int argc, char **agrv)
{
#ifdef LINK_LIST_TEST
        cout << "-------- Begin Link List tests --------" << endl;
        link_list_test(); // test link list
        cout << "-------- End Link List tests --------" << endl;
#endif
        return 1;
}
